chore(website): update build script to use esbuild instead of rollup

armano2 opened the pull request:
## PR Checklist

- [x] Addresses an existing open issue: [#6656 comment](https://github.com/typescript-eslint/typescript-eslint/pull/6656#pullrequestreview-1347474936)
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

those changes have been extracted from #6656 

- migrate from rollup to esbuild
- correct source map generation
- reduce bundle size
- simplify build

### stats
- bundling website-eslint: 220.45ms
- netlify build time: ~ -2m (if we remove building all packages)
- package size: ~ -250kb


<details>

<summary>source maps</summary>

#### this branch
![image](https://user-images.githubusercontent.com/625469/227501689-e9903561-f9b6-4c67-88a5-04b4cf772b14.png)

#### v6
![image](https://user-images.githubusercontent.com/625469/227501865-dfa8704c-e13b-4def-a46d-8329d950f6b3.png)

</details>

https://github.com/evanw/esbuild/issues/819

bradzacher commented:
ditto with https://github.com/typescript-eslint/typescript-eslint/pull/6656 - could we base this against the v6 branch?

armano2 commented on yarn.lock:
```diff
@@ -6597,6 +6637,34 @@ esbuild@~0.15.10:
     esbuild-windows-64 "0.15.18"
     esbuild-windows-arm64 "0.15.18"
 
+esbuild@~0.17.12:
```
some changes from main are not merged yet, and we have older version of esbuild used in other places

armano2 commented on packages/website-eslint/src/mock/path.js:
```diff
@@ -212,11 +212,11 @@ export function basename(path, ext) {
 }
 
-export default {
+module.exports = {
```
all packages that we are bundling are cjs, and there is significantly less code generated if format of all packages is the same, as there is no need to convert between cjs/esm exports/imports

armano2 replied:
left: mixed esm, right this branch (some of differences)

![image](https://user-images.githubusercontent.com/625469/226686751-a696480b-f12e-411f-bd20-5e41875b185f.png)

![image](https://user-images.githubusercontent.com/625469/226686852-4f4fad7a-a3ec-442d-97e9-b26849ada962.png)

![image](https://user-images.githubusercontent.com/625469/226687463-3787a57e-3e4f-4bce-adf4-350bd65b63fc.png)

![image](https://user-images.githubusercontent.com/625469/226687863-fbdbe4f4-8a30-47cd-92c1-eaa48b75090f.png)

------------

esm is generally way better, but not in mixed content

JoshuaKGoldberg replied:
Ooh, how are you generating these visualizations? I love them 😍

armano2 replied:
i turned of minimize in build, generated file, copied it, as index2.js

and compared in my editor with generated code with changes,

I'm not using vscode tho, its intelij

![image](https://user-images.githubusercontent.com/625469/226810035-78b703c8-57a3-44f0-8580-8a1fa8cb64a7.png)

armano2 replied:
i started thinking about this, and i have one idea, maybe instead of building everything from compiled code we could build it from sources?

armano2 commented on packages/typescript-estree/package.json:
```diff
@@ -15,17 +15,9 @@
-    "./use-at-your-own-risk/getScriptKind": {
-      "types": "./dist/create-program/getScriptKind.d.ts",
-      "default": "./dist/create-program/getScriptKind.js"
+    "./use-at-your-own-risk": {
+      "types": "./dist/use-at-your-own-risk.d.ts",
+      "default": "./dist/use-at-your-own-risk.js"
```
i'm not sure about this name

armano2 commented:
@bradzacher i enabled type-checking for `website-eslint` and updated build to use source code instead of `dist`, I also removed most of no longer necessary `use-at-your-own-risk` as some of them where redundant and not really necessary

if you think that we should restore some of them let me know

armano2 commented on packages/website-eslint/build.ts:
```diff
@@ -0,0 +1,167 @@
+        setup(build): void {
+          build.onLoad(
+            makeFilter([
+              '/eslint-utils/rule-tester/RuleTester.ts',
```
we should fix this, this file is imported only as type but has non type dependencies

bradzacher reviewed (approved):
remove the rollup config from renovate and we can land this

https://github.com/typescript-eslint/typescript-eslint/blob/8ee0cbb0818392cc818b1c254dcec6e879f79a35/.github/renovate.json5#L19-L28

https://github.com/typescript-eslint/typescript-eslint/blob/8ee0cbb0818392cc818b1c254dcec6e879f79a35/.github/renovate.json5#L88-L92
